tools/xkbcli-scaffold-new-layout.py

```diff
@@ -0,0 +1,323 @@
+    for d in ['compat', 'keycodes', 'rules', 'symbols', 'types']:
+        (basedir / d).mkdir(exist_ok=True)
+
+
+def create_rules_template(base_path, ruleset, layout, option):
+    rules = Path(base_path) / 'rules' / ruleset
+    if rules.exists():
+        logger.warning(f'Rules file {rules} already exists, skipping')
+        return
+
+    with open(rules, 'w') as fd:
+        header = dedent(f'''\
+            // generated by xkbcli scaffold-new-layout
+
+            // Note: no rules file entries are required for for a custom layout
+
+            ''')
+        fd.write(header)
+
+        if option:
+            group, section = option
+            option_template = dedent(f'''\
+            // This section maps XKB option "{group}:{section}" to the '{section}' section in the
+            // 'symbols/{group}' file.
+            //
+            ! option = symbols
+              {group}:{section} = +{group}({section})
+
+            ''')
+            fd.write(option_template)
+
+        footer = dedent(f'''\
+            // Include the system '{ruleset}' file
+            ! include %S/{ruleset}
+            ''')
+        fd.write(footer)
+
+
+def create_symbols_template(basedir, layout_variant, option):
+    if not layout_variant and not option:
+        logger.info('No layout or option given, skipping symbols templates')
+        return
+
+    layout, variant = layout_variant
+    layout_file = Path(basedir) / 'symbols' / layout
+    if layout_file.exists():
+        logger.warning(f'Symbols file {layout_file} already exists, skipping')
+        layout_fd = None
+    else:
+        layout_fd = open(layout_file, 'w')
+        layout_fd.write('// generated by xkbcli scaffold-new-sources\n\n')
+
+    group, section = option
+    options_file = Path(basedir) / 'symbols' / group
+
+    # Cater for a potential "custom(variant)" layout and "custom:foo" option,
+    # i.e. where both layout and options use the same symbols file
+    if options_file == layout_file:
+        options_fd = layout_fd
+    elif options_file.exists():
+        logger.warning(f'File {options_file} already exists, skipping')
+        options_fd = None
+    else:
+        options_fd = open(options_file, 'w')
+        options_fd.write('// generated by xkbcli scaffold\n\n')
+
+    if layout_fd:
+        if variant is None:
+            default = 'default '
+            variant = 'basic'
+            include = ''
+        else:
+            default = ''
+            include = f'include "{layout}(basic)"'
+
+        logger.debug(f'Writing "{layout}({variant})" layout template to {layout_file}')
+        layout_fd.write(dedent(f'''\
+                {default}partial alphanumeric_keys modifier_keys
+                xkb_symbols "{variant}" {{
+                    name[Group1]= "{variant} ({layout})";
+
+                    {include}
+
+                    // Example:
+                    // key <CAPS> {{ [ Escape ] }};
+                }};
+                '''))
```

Comment:
Should close `layout_fd` here (same for `options_fd` below)

Reply:
`layout_fd` and `options_fd` may point to the same file so we can't close them until the end of the file. I've added that now, thanks.